protocols/smtp/src/test/java/org/apache/james/protocols/smtp/AbstractSMTPServerTest.java

```diff
@@ -69,43 +73,119 @@ public void testSimpleDelivery() throws Exception {
             server = createServer(createProtocol(hook));
             server.bind();
             
-            SMTPClient client = createClient();
             InetSocketAddress bindedAddress = new ProtocolServerUtils(server).retrieveBindedAddress();
-            client.connect(bindedAddress.getAddress().getHostAddress(), bindedAddress.getPort());
-            assertTrue(SMTPReply.isPositiveCompletion(client.getReplyCode()));
-            
-            client.helo("localhost");
-            assertTrue(SMTPReply.isPositiveCompletion(client.getReplyCode()));
 
-            client.setSender(SENDER);
-            assertTrue("Reply="+ client.getReplyString(), SMTPReply.isPositiveCompletion(client.getReplyCode()));
+            send(server, bindedAddress, MSG1);
 
-            client.addRecipient(RCPT1);
-            assertTrue("Reply="+ client.getReplyString(), SMTPReply.isPositiveCompletion(client.getReplyCode()));
+            Iterator<MailEnvelope> queued = hook.getQueued().iterator();
+            assertTrue(queued.hasNext());
+            
+            MailEnvelope env = queued.next();
+            checkEnvelope(env, SENDER, Arrays.asList(RCPT1, RCPT2), MSG1);
+            assertFalse(queued.hasNext());
 
-            client.addRecipient(RCPT2);
-            assertTrue("Reply="+ client.getReplyString(), SMTPReply.isPositiveCompletion(client.getReplyCode()));
+        } finally {
+            if (server != null) {
+                server.unbind();
+            }
+        }
+        
+    }
+    
+    @Test
+    public void testDeliveryWith4SimultaneousThreads() throws Exception {
+        TestMessageHook hook = new TestMessageHook();
+        ProtocolServer server = null;
+        try {
+            server = createServer(createProtocol(hook));
+            server.bind();
+            InetSocketAddress bindedAddress = new ProtocolServerUtils(server).retrieveBindedAddress();
 
-            assertTrue(client.sendShortMessageData(MSG1));
-            assertTrue("Reply="+ client.getReplyString(), SMTPReply.isPositiveCompletion(client.getReplyCode()));
-            
-            client.quit();
-            assertTrue("Reply="+ client.getReplyString(), SMTPReply.isPositiveCompletion(client.getReplyCode()));
-            client.disconnect();
+            String mailContent = CharStreams.toString(new InputStreamReader(ClassLoader.getSystemResourceAsStream("a50.eml"), Charsets.US_ASCII));
+            Thread thread1 = new SendThread(server, bindedAddress, mailContent);
+            Thread thread2 = new SendThread(server, bindedAddress, mailContent);
+            Thread thread3 = new SendThread(server, bindedAddress, mailContent);
+            Thread thread4 = new SendThread(server, bindedAddress, mailContent);
+            thread1.start();

```

Comment:
Couldn't you use the **ConcurrentTestRunner** that I recently introduced ?

Seems like to be the tool you are looking for...

You can factorize it in java-utils.

It will : 
 - Allow to have a tested stuff for running parrallel tests
 - Have a tool better than threads (ThreadPool)
 - better synchronization than chaining calls to start
 - Reduce the code size
 - I believe make it more readable.